Python.h not found during cross-compilation

When cross-compiling python3-cryptography v40.0.2 with Yocto it fails if you have different version of Python on your host and in the Yocto build. This is caused by `src/rust/build.rs` that is executing a python script to find the location of Python.h.

Reference: https://lists.openembedded.org/g/openembedded-core/topic/98667212 

```
|   running: "x86_64-poky-linux-gcc" "-m64" "-march=core2" "-mtune=core2" "-msse3" "-mfpmath=sse" "-fstack-protector-strong" "-O2" "-D_FORTIFY_SOURCE=2" "-Wformat" "-Wformat-security" "-Werror=format-security" "--sysroot=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot" "-O2" "-pipe" "-g" "-feliminate-unused-debug-types" "-fmacro-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fdebug-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fmacro-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fdebug-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fdebug-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=" "-fmacro-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=" "-fdebug-prefix-map=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native=" "-I" "/usr/include/python3.11" "-I" "/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot/usr/include" "-Wconversion" "-Wno-error=sign-conversion" "-Wno-unused-parameter" "-DPy_LIMITED_API=0x030600f0" "-o" "/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/x86_64-poky-linux-gnu/release/build/cryptography-rust-aa00e39a952c07ee/out/9f543b84f9f4dceb-_openssl.o" "-c" "/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/x86_64-poky-linux-gnu/release/build/cryptography-rust-aa00e39a952c07ee/out/_openssl.c"
|   cargo:warning=/srv/build/ttorling/build-upgrades/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/x86_64-poky-linux-gnu/release/build/cryptography-rust-aa00e39a952c07ee/out/_openssl.c:57:10: fatal error: Python.h: No such file or directory
|   cargo:warning=   57 | #include <Python.h>
|   cargo:warning=      |          ^~~~~~~~~~
|   cargo:warning=compilation terminated.
|   exit status: 1
```

The setup in Yocto is that you have one sysroot for the native tools used to build things for the target system and one sysroot for target with target files. In my particular setup with Ubuntu 22.04 I have python3.10 on my machine. Yocto's native tools have python3.11. When running https://github.com/pyca/cryptography/blob/f816b457494e010b655cd7fdcd30e3446f86a703/src/rust/build.rs#L46 the answer from Yocto will be `/usr/include/python3.11` that is then passed to the crooss-compilaton build and it fails to find Python.h. The weak part here is that the code is running python on in a native environment to figure out the include path for the cross-compilation.

In Yocto this can be solved and handled by using internal variable instead but raising this here if we could find a better upstream solution to avoid local patching in Yocto. Hopefully this can also be good for others that will cross-compile this project. Is it possible to use `pkg-config` or other way instead to find location of header files?

> However, if someone can help me understand the answer to:
> 
> > I guess what I don't understand is how this worked under the previous setup! What was causing a host python to compile+link against a set of headers+libraries besides its own?
> 
> then perhaps we can find a workable solution here.

> Be that as it may, we're not going to fix fundamental issues in Python packaging today :-)
> 
> However, if someone can help me understand the answer to:
> 
> > I guess what I don't understand is how this worked under the previous setup! What was causing a host python to compile+link against a set of headers+libraries besides its own?
> 
> then perhaps we can find a workable solution here.

I haven't found all the details yet but start with what I found and we can hopefully fill in the unknowns together.

A common start for the old and new versions in Yocto are:

`/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native/usr/bin/python3-native/python3 setup.py bdist_wheel --verbose --dist-dir /work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/dist`

python-cryptography_39.0.2:
```
building 'cryptography.hazmat.bindings._openssl' extension
creating build/temp.linux-x86_64-cpython-311/build
creating build/temp.linux-x86_64-cpython-311/build/temp.linux-x86_64-cpython-311
x86_64-poky-linux-gcc -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse -fstack-protector-strong -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot -Wsign-compare -DNDEBUG -g -O3 -Wall -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemux86-64-systemd/tmp/work/core2-64-poky-linux/python3/3.11.2-r0/Python-3.11.2=/usr/src/debug/python3/3.11.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemux86-64-systemd/tmp/work/core2-64-poky-linux/python3/3.11.2-r0/Python-3.11.2=/usr/src/debug/python3/3.11.2-r0 -fmacro-prefix-map=/build/path/unavailable/=/usr/src/debug/python3/3.11.2-r0 -fdebug-prefix-map=/build/path/unavailable/=/usr/src/debug/python3/3.11.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot= -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot= -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot-native= -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemux86-64-systemd/tmp/work/core2-64-poky-linux/python3/3.11.2-r0/Python-3.11.2=/usr/src/debug/python3/3.11.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemux86-64-systemd/tmp/work/core2-64-poky-linux/python3/3.11.2-r0/Python-3.11.2=/usr/src/debug/python3/3.11.2-r0 -fmacro-prefix-map=/build/path/unavailable/=/usr/src/debug/python3/3.11.2-r0 -fdebug-prefix-map=/build/path/unavailable/=/usr/src/debug/python3/3.11.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot= -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot= -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot-native= -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/cryptography-39.0.2=/usr/src/debug/python3-cryptography-old/39.0.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/cryptography-39.0.2=/usr/src/debug/python3-cryptography-old/39.0.2-r0 -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/build=/usr/src/debug/python3-cryptography-old/39.0.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/build=/usr/src/debug/python3-cryptography-old/39.0.2-r0 -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot= -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot= -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot-native= -fPIC -I/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot/usr/include/python3.11 -c build/temp.linux-x86_64-cpython-311/cryptography.hazmat.bindings._openssl.c -o build/temp.linux-x86_64-cpython-311/build/temp.linux-x86_64-cpython-311/cryptography.hazmat.bindings._openssl.o -Wconversion -Wno-error=sign-conversion
```

This will add `-I/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography-old/39.0.2-r0/recipe-sysroot/usr/include/python3.11` to the compilation.

python-cryptography_40.0.0:

```
error: failed to run custom build command for `cryptography-rust v0.1.0 (/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2/src/rust)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_RELEASE_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/release/build/cryptography-rust-5fcf0969965df7b9/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=../_cffi_src/
  cargo:rustc-cfg=python_implementation="CPython"
  TARGET = Some("x86_64-poky-linux-gnu")
  OPT_LEVEL = Some("3")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_x86_64-poky-linux-gnu
  CC_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CC_x86_64_poky_linux_gnu
  CC_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = Some("x86_64-poky-linux-gcc  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot")
  cargo:rerun-if-env-changed=CFLAGS_x86_64-poky-linux-gnu
  CFLAGS_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_poky_linux_gnu
  CFLAGS_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = Some(" -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native= ")
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = Some("1")
  Info: default compiler flags are disabled
  cargo:rerun-if-env-changed=CC_x86_64-poky-linux-gnu
  CC_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CC_x86_64_poky_linux_gnu
  CC_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = Some("x86_64-poky-linux-gcc  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot")
  cargo:rerun-if-env-changed=CFLAGS_x86_64-poky-linux-gnu
  CFLAGS_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_poky_linux_gnu
  CFLAGS_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = Some(" -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native= ")
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = Some("1")
  Info: default compiler flags are disabled
  cargo:rerun-if-env-changed=CC_x86_64-poky-linux-gnu
  CC_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CC_x86_64_poky_linux_gnu
  CC_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = Some("x86_64-poky-linux-gcc  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot")
  cargo:rerun-if-env-changed=CFLAGS_x86_64-poky-linux-gnu
  CFLAGS_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_poky_linux_gnu
  CFLAGS_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = Some(" -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native= ")
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = Some("1")
  Info: default compiler flags are disabled
  cargo:rerun-if-env-changed=CC_x86_64-poky-linux-gnu
  CC_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CC_x86_64_poky_linux_gnu
  CC_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = Some("x86_64-poky-linux-gcc  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot")
  cargo:rerun-if-env-changed=CFLAGS_x86_64-poky-linux-gnu
  CFLAGS_x86_64-poky-linux-gnu = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_poky_linux_gnu
  CFLAGS_x86_64_poky_linux_gnu = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = Some(" -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=  -fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native= ")
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = Some("1")
  Info: default compiler flags are disabled
  running: "x86_64-poky-linux-gcc" "-m64" "-march=core2" "-mtune=core2" "-msse3" "-mfpmath=sse" "-fstack-protector-strong" "-O2" "-D_FORTIFY_SOURCE=2" "-Wformat" "-Wformat-security" "-Werror=format-security" "--sysroot=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot" "-O2" "-pipe" "-g" "-feliminate-unused-debug-types" "-fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/cryptography-40.0.2=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build=/usr/src/debug/python3-cryptography/40.0.2-r0" "-fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=" "-fmacro-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot=" "-fdebug-prefix-map=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot-native=" "-I" "/usr/include/python3.11" "-I" "/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/recipe-sysroot/usr/include" "-Wconversion" "-Wno-error=sign-conversion" "-Wno-unused-parameter" "-DPy_LIMITED_API=0x030600f0" "-o" "/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/x86_64-poky-linux-gnu/release/build/cryptography-rust-692a1fa095ab554f/out/79a447d161ca18d2-_openssl.o" "-c" "/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/x86_64-poky-linux-gnu/release/build/cryptography-rust-692a1fa095ab554f/out/_openssl.c"
  cargo:warning=/work/yocto/poky/build/build-qemuarm64-systemd/tmp/work/core2-64-poky-linux/python3-cryptography/40.0.2-r0/build/target/x86_64-poky-linux-gnu/release/build/cryptography-rust-692a1fa095ab554f/out/_openssl.c:57:10: fatal error: Python.h: No such file or directory
  cargo:warning=   57 | #include <Python.h>
  cargo:warning=      |          ^~~~~~~~~~
  cargo:warning=compilation terminated.
  exit status: 1
```
This will add `"-I" "/usr/include/python3.11"` to the compilation.

One difference is that version 40 is building inside a rust context while version 39 is not doing that. I haven't found how the 'good' path is passed to the build in version 39.

I'll definitely submit a PR with the above, I'm just trying to make sense of a different, seemingly unrelated fail that happens later in the build, and is as well new to 40 and later:
```
  = note: /srv/storage/alex/yocto/build-64-alt/tmp/work/core2-64-poky-linux/python3-cryptography/41.0.1-r0/recipe-sysroot-native/usr/bin/x86_64-poky-linux/../../libexec/x86_64-poky-linux/gcc/x86_64-poky-linux/13
.1.1/ld: /srv/storage/alex/yocto/build-64-alt/tmp/work/core2-64-poky-linux/python3-cryptography/41.0.1-r0/build/target/x86_64-poky-linux-gnu/release/deps/libcryptography_cffi-74a91ece88c27f69.rlib(10a8fe5b03ab2d
a5-_openssl.o): warning: relocation against `_Py_NoneStruct' in read-only section `.text'
          /srv/storage/alex/yocto/build-64-alt/tmp/work/core2-64-poky-linux/python3-cryptography/41.0.1-r0/recipe-sysroot-native/usr/bin/x86_64-poky-linux/../../libexec/x86_64-poky-linux/gcc/x86_64-poky-linux/13
.1.1/ld: /srv/storage/alex/yocto/build-64-alt/tmp/work/core2-64-poky-linux/python3-cryptography/41.0.1-r0/build/target/x86_64-poky-linux-gnu/release/deps/libcryptography_cffi-74a91ece88c27f69.rlib(10a8fe5b03ab2d
a5-_openssl.o): relocation R_X86_64_PC32 against undefined symbol `_Py_NoneStruct' can not be used when making a shared object; recompile with -fPIC
          /srv/storage/alex/yocto/build-64-alt/tmp/work/core2-64-poky-linux/python3-cryptography/41.0.1-r0/recipe-sysroot-native/usr/bin/x86_64-poky-linux/../../libexec/x86_64-poky-linux/gcc/x86_64-poky-linux/13
.1.1/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status
          

error: could not compile `cryptography-rust` (lib) due to previous error
```

For poky users, work-in-progress 39.0.2 -> 41.0.1 update can be found here:
https://git.yoctoproject.org/poky-contrib/log/?h=akanavin/package-version-updates

--- a/src/rust/cryptography-cffi/build.rs
+++ b/src/rust/cryptography-cffi/build.rs
@@ -49,7 +49,7 @@ fn main() {
     println!("cargo:rustc-cfg=python_implementation=\"{}\"", python_impl);
     let python_include = run_python_script(
         &python,
-        "import sysconfig; print(sysconfig.get_path('include'), end='')",
+        "import sysconfig; print(sysconfig.get_config_var('INCLUDEPY'), end='')",
     )
     .unwrap();
     let openssl_include =

